internal/rotate_events.go

```diff
@@ -0,0 +1,94 @@
+package internal
+
+import (
+	"log"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// EventSync in an interface for sending events to device-gateway. The abstraction
+// makes it easier to write unit tests
+type EventSync interface {
+	NotifyStarted()
+	NotifyStep(name string, err error)
+	NotifyCompleted(err error)
+	SetCorrelationId(corId string)
+}
+
+type NoOpEventSync struct{}
+
+func (s NoOpEventSync) NotifyStarted()                    {}
+func (s NoOpEventSync) NotifyCompleted(err error)         {}
+func (s NoOpEventSync) NotifyStep(name string, err error) {}
+func (s NoOpEventSync) SetCorrelationId(corId string)     {}
+
+type DgEventSync struct {
+	client        *http.Client
+	url           string
+	correlationId string
+	target        CurrentTarget
+}
+
+func (s *DgEventSync) SetCorrelationId(corId string) {
+	s.correlationId = corId
+}
+func (s *DgEventSync) NotifyStarted() {
+	s.notify("CertRotationStarted", nil)
+}
+func (s *DgEventSync) NotifyCompleted(err error) {
+	s.notify("CertRotationCompleted", err)
+}
+func (s *DgEventSync) NotifyStep(name string, err error) {
+	s.notify(name, err)
+}
+func (s *DgEventSync) notify(event string, err error) {
+	details := ""
+	if err != nil {
+		details = err.Error()
+	}
+	evt := []DgUpdateEvent{
+		{
+			Id:         uuid.New().String(),
+			DeviceTime: time.Now().Format(time.RFC3339),
+			Event: DgEvent{
+				CorrelationId: s.correlationId,
+				Success:       err == nil,
+				TargetName:    s.target.Name,
+				Version:       strconv.Itoa(s.target.Version),
```

Comment:
Generally speaking I'm not sure we need to bother with the TargetName and Version for the certificate rotation.
These are not the target update events; so IMHO target-related fields provides no value here.
IIRC we don't rely onto these two fields anywhere in our stack (that is expected) so we should be fine here, right?

That said, I have no strong opinion on this; except for the desire to support less code here.

Reply:
i started w/o them. It looks bad in the UI. It's also kind of handy when looking at update events to see "this rotation happened when the device was running version $X".

Reply:
I'm fine with it; it was more a question.

If an added value overcomes the added maintenance cost, go for it.